File: lib/data/services/robot_service.dart
import 'dart:async';
import 'package:dio/dio.dart';
import 'package:logger/logger.dart';

/// Dispositivo Robot
class RobotDevice {

  const RobotDevice({
    required this.id,
    required this.name,
    required this.model,
    required this.serialNumber,
    required this.firmwareVersion,
    required this.bluetoothId,
    required this.status,
    required this.lastSeen,
    required this.createdAt,
    required this.updatedAt,
  });

  factory RobotDevice.fromJson(Map<String, dynamic> json) =>
      RobotDevice(
        id: json['id'] as String,
        name: json['name'] as String,
        model: json['model'] as String,
        serialNumber: json['serialNumber'] as String,
        firmwareVersion: json['firmwareVersion'] as String,
        bluetoothId: json['bluetoothId'] as String,
        status: json['status'] as String,
        lastSeen: DateTime.parse(json['lastSeen'] as String),
        createdAt: DateTime.parse(json['createdAt'] as String),
        updatedAt: DateTime.parse(json['updatedAt'] as String),
      );
  final String id;
  final String name;
  final String model;
  final String serialNumber;
  final String firmwareVersion;
  final String bluetoothId;
  final String status; // 'online' | 'offline' | 'setup' | 'error'
  final DateTime lastSeen;
  final DateTime createdAt;
  final DateTime updatedAt;

  Map<String, dynamic> toJson() => {
    'id': id,
    'name': name,
    'model': model,
    'serialNumber': serialNumber,
    'firmwareVersion': firmwareVersion,
    'bluetoothId': bluetoothId,
    'status': status,
    'lastSeen': lastSeen.toIso8601String(),
    'createdAt': createdAt.toIso8601String(),
    'updatedAt': updatedAt.toIso8601String(),
  };
}

/// Solicitud de validación del robot
class RobotValidationRequest {

  const RobotValidationRequest({
    required this.deviceId,
    required this.deviceName,
    required this.bluetoothId,
    this.model,
    this.serialNumber,
  });
  final String deviceId;
  final String deviceName;
  final String bluetoothId;
  final String? model;
  final String? serialNumber;

  Map<String, dynamic> toJson() => {
    'deviceId': deviceId,
    'deviceName': deviceName,
    'bluetoothId': bluetoothId,
    'model': model,
    'serialNumber': serialNumber,
  };
}

/// Respuesta de validación del robot
class RobotValidationResponse {

  const RobotValidationResponse({
    required this.isValid,
    required this.message, this.device,
    this.requiresUpdate = false,
  });

  factory RobotValidationResponse.fromJson(Map<String, dynamic> json) =>
      RobotValidationResponse(
        isValid: json['isValid'] as bool,
        device: json['device'] != null 
            ? RobotDevice.fromJson(json['device']) 
            : null,
        message: json['message'] as String,
        requiresUpdate: json['requiresUpdate'] as bool? ?? false,
      );
  final bool isValid;
  final RobotDevice? device;
  final String message;
  final bool requiresUpdate;
}

/// Solicitud de configuración WiFi
class WiFiConfigurationRequest {

  const WiFiConfigurationRequest({
    required this.deviceId,
    required this.wifiSSID,
    required this.wifiPassword,
    this.encryptionType,
  });
  final String deviceId;
  final String wifiSSID;
  final String wifiPassword;
  final String? encryptionType;

  Map<String, dynamic> toJson() => {
    'deviceId': deviceId,
    'wifiSSID': wifiSSID,
    'wifiPassword': wifiPassword,
    'encryptionType': encryptionType,
  };
}

/// Respuesta de configuración WiFi
class WiFiConfigurationResponse {

  const WiFiConfigurationResponse({
    required this.success,
    required this.message,
    this.connectionTest,
  });

  factory WiFiConfigurationResponse.fromJson(Map<String, dynamic> json) =>
      WiFiConfigurationResponse(
        success: json['success'] as bool,
        message: json['message'] as String,
        connectionTest: json['connectionTest'] != null
            ? ConnectionTest.fromJson(json['connectionTest'])
            : null,
      );
  final bool success;
  final String message;
  final ConnectionTest? connectionTest;
}

/// Prueba de conexión
class ConnectionTest {

  const ConnectionTest({
    required this.success,
    required this.ping,
    required this.downloadSpeed,
    required this.uploadSpeed,
  });

  factory ConnectionTest.fromJson(Map<String, dynamic> json) =>
      ConnectionTest(
        success: json['success'] as bool,
        ping: (json['ping'] as num).toDouble(),
        downloadSpeed: (json['downloadSpeed'] as num).toDouble(),
        uploadSpeed: (json['uploadSpeed'] as num).toDouble(),
      );
  final bool success;
  final double ping;
  final double downloadSpeed;
  final double uploadSpeed;
}

/// Comando del robot
class RobotCommand {

  const RobotCommand({
    required this.deviceId,
    required this.command,
    required this.parameters,
    required this.timestamp,
  });
  final String deviceId;
  final String command;
  final Map<String, dynamic> parameters;
  final DateTime timestamp;

  Map<String, dynamic> toJson() => {
    'deviceId': deviceId,
    'command': command,
    'parameters': parameters,
    'timestamp': timestamp.toIso8601String(),
  };
}

/// Servicio de Robot
class RobotService {

  RobotService({
    required Logger logger,
    required Dio dio,
  }) : _logger = logger,
       _dio = dio;
  final Logger _logger;
  final Dio _dio;

  final List<RobotDevice> _devices = [];
  
  // Streams para notificaciones
  final StreamController<List<RobotDevice>> _devicesController = 
      StreamController<List<RobotDevice>>.broadcast();
  final StreamController<RobotDevice> _deviceStatusController = 
      StreamController<RobotDevice>.broadcast();

  /// Validar dispositivo robot
  Future<RobotValidationResponse> validateDevice(RobotValidationRequest request) async {
    try {
      _logger.i('Validating robot device: ${request.deviceName}');

      final response = await _dio.post(
        '/robots/validate',
        data: request.toJson(),
      );

      final validationResponse = RobotValidationResponse.fromJson(response.data);
      
      if (validationResponse.isValid && validationResponse.device != null) {
        _updateDeviceInList(validationResponse.device!);
      }

      _logger.i('Device validation result: ${validationResponse.isValid}');
      return validationResponse;
    } catch (e) {
      _logger.e('Error validating robot device: $e');
      
      // Simular respuesta de error para desarrollo
      return RobotValidationResponse(
        isValid: false,
        message: 'Validation failed: $e',
      );
    }
  }

  /// Obtener dispositivos robot del usuario
  Future<List<RobotDevice>> getUserDevices() async {
    try {
      _logger.i('Fetching user robot devices');

      final response = await _dio.get('/robots');

      final devicesJson = response.data['devices'] as List;
      final devices = devicesJson
          .map((json) => RobotDevice.fromJson(json))
          .toList();

      _devices.clear();
      _devices.addAll(devices);
      _devicesController.add(List.unmodifiable(_devices));

      _logger.i('Found ${devices.length} robot devices');
      return devices;
    } catch (e) {
      _logger.e('Error fetching robot devices: $e');
      
      // Simular datos de prueba para desarrollo
      final mockDevices = <RobotDevice>[
        RobotDevice(
          id: 'robot-1',
          name: 'Nebu Robot Alpha',
          model: 'NR-1000',
          serialNumber: 'NR1000-001',
          firmwareVersion: '1.2.3',
          bluetoothId: '00:11:22:33:44:55',
          status: 'online',
          lastSeen: DateTime.now().subtract(const Duration(minutes: 5)),
          createdAt: DateTime.now().subtract(const Duration(days: 30)),
          updatedAt: DateTime.now().subtract(const Duration(hours: 2)),
        ),
        RobotDevice(
          id: 'robot-2',
          name: 'Nebu Robot Beta',
          model: 'NR-2000',
          serialNumber: 'NR2000-002',
          firmwareVersion: '2.1.0',
          bluetoothId: '00:11:22:33:44:66',
          status: 'offline',
          lastSeen: DateTime.now().subtract(const Duration(hours: 2)),
          createdAt: DateTime.now().subtract(const Duration(days: 15)),
          updatedAt: DateTime.now().subtract(const Duration(hours: 2)),
        ),
      ];

      _devices.clear();
      _devices.addAll(mockDevices);
      _devicesController.add(List.unmodifiable(_devices));

      return mockDevices;
    }
  }

  /// Configurar WiFi en el robot
  Future<WiFiConfigurationResponse> configureWiFi(WiFiConfigurationRequest request) async {
    try {
      _logger.i('Configuring WiFi for robot: ${request.deviceId}');

      final response = await _dio.post(
        '/robots/${request.deviceId}/wifi',
        data: request.toJson(),
      );

      final configResponse = WiFiConfigurationResponse.fromJson(response.data);
      
      _logger.i('WiFi configuration result: ${configResponse.success}');
      return configResponse;
    } catch (e) {
      _logger.e('Error configuring WiFi: $e');
      
      // Simular respuesta para desarrollo
      return WiFiConfigurationResponse(
        success: false,
        message: 'WiFi configuration failed: $e',
      );
    }
  }

  /// Enviar comando al robot
  Future<bool> sendCommand(RobotCommand command) async {
    try {
      _logger.i('Sending command to robot: ${command.deviceId} - ${command.command}');

      await _dio.post(
        '/robots/${command.deviceId}/commands',
        data: command.toJson(),
      );

      _logger.i('Command sent successfully');
      return true;
    } catch (e) {
      _logger.e('Error sending command to robot: $e');
      return false;
    }
  }

  /// Mover robot
  Future<bool> moveRobot({
    required String deviceId,
    required String direction,
    required double speed,
    required Duration duration,
  }) async {
    final command = RobotCommand(
      deviceId: deviceId,
      command: 'move',
      parameters: {
        'direction': direction,
        'speed': speed,
        'duration': duration.inMilliseconds,
      },
      timestamp: DateTime.now(),
    );

    return sendCommand(command);
  }

  /// Detener robot
  Future<bool> stopRobot(String deviceId) async {
    final command = RobotCommand(
      deviceId: deviceId,
      command: 'stop',
      parameters: {},
      timestamp: DateTime.now(),
    );

    return sendCommand(command);
  }

  /// Actualizar firmware del robot
  Future<bool> updateFirmware(String deviceId) async {
    try {
      _logger.i('Updating firmware for robot: $deviceId');

      await _dio.post('/robots/$deviceId/firmware/update');
      
      _logger.i('Firmware update initiated');
      return true;
    } catch (e) {
      _logger.e('Error updating firmware: $e');
      return false;
    }
  }

  /// Obtener estado del robot
  Future<RobotDevice?> getDeviceStatus(String deviceId) async {
    try {
      final response = await _dio.get('/robots/$deviceId/status');
      
      final device = RobotDevice.fromJson(response.data);
      _updateDeviceInList(device);
      
      return device;
    } catch (e) {
      _logger.e('Error getting device status: $e');
      return null;
    }
  }

  /// Actualizar dispositivo en la lista
  void _updateDeviceInList(RobotDevice device) {
    final index = _devices.indexWhere((d) => d.id == device.id);
    
    if (index != -1) {
      _devices[index] = device;
    } else {
      _devices.add(device);
    }
    
    _devicesController.add(List.unmodifiable(_devices));
    _deviceStatusController.add(device);
  }

  /// Obtener dispositivos actuales
  List<RobotDevice> get devices => List.unmodifiable(_devices);

  /// Obtener dispositivo por ID
  RobotDevice? getDeviceById(String deviceId) {
    try {
      return _devices.firstWhere((device) => device.id == deviceId);
    } catch (e) {
      return null;
    }
  }

  /// Stream de dispositivos
  Stream<List<RobotDevice>> get devicesStream => _devicesController.stream;

  /// Stream de estado de dispositivo
  Stream<RobotDevice> get deviceStatusStream => _deviceStatusController.stream;

  /// Cerrar servicio
  Future<void> dispose() async {
    await _devicesController.close();
    await _deviceStatusController.close();
    _logger.i('Robot Service disposed');
  }
}

File: backup_services/robotService.ts
import { logger } from '@/utils/logger';
import { apiService } from './api';

// Tipos para el servicio del robot
export interface RobotDevice {
  id: string;
  name: string;
  model: string;
  serialNumber: string;
  firmwareVersion: string;
  bluetoothId: string;
  status: 'online' | 'offline' | 'setup' | 'error';
  lastSeen: string;
  createdAt: string;
  updatedAt: string;
}

export interface RobotValidationRequest {
  deviceId: string;
  deviceName: string;
  bluetoothId: string;
  model?: string;
  serialNumber?: string;
}

export interface RobotValidationResponse {
  isValid: boolean;
  device?: RobotDevice;
  message: string;
  requiresUpdate?: boolean;
}

export interface WiFiConfigurationRequest {
  deviceId: string;
  wifiSSID: string;
  wifiPassword: string;
  encryptionType?: string;
}

export interface WiFiConfigurationResponse {
  success: boolean;
  message: string;
  connectionTest?: {
    success: boolean;
    ping: number;
    downloadSpeed: number;
    uploadSpeed: number;
  };
}

export interface RobotStatus {
  deviceId: string;
  status: 'online' | 'offline' | 'setup' | 'error';
  batteryLevel?: number;
  signalStrength?: number;
  lastCommand?: string;
  lastCommandTime?: string;
  errors?: string[];
}

class RobotService {
  private baseEndpoint = '/iot/devices';

  /**
   * Validar que un robot existe en el backend
   */
  async validateRobot(request: RobotValidationRequest): Promise<RobotValidationResponse> {
    try {
      logger.debug('Validating robot in backend:', request);

      // Llamada real al backend
      const response = await apiService.post(`${this.baseEndpoint}/validate`, request);
      
      return {
        isValid: response.data.isValid,
        device: response.data.device,
        message: response.data.message,
        requiresUpdate: response.data.requiresUpdate || false,
      };
    } catch (error) {
      logger.error('Error validating robot:', error);
      
      return {
        isValid: false,
        message: 'Failed to validate robot. Please check your internet connection.',
      };
    }
  }

  /**
   * Configurar WiFi en el robot
   */
  async configureWiFi(request: WiFiConfigurationRequest): Promise<WiFiConfigurationResponse> {
    try {
      logger.debug('Configuring WiFi for robot:', request.deviceId);

      // Simular llamada al backend
      await new Promise(resolve => setTimeout(resolve, 3000));

      // Simular respuesta del backend
      const mockResponse: WiFiConfigurationResponse = {
        success: true,
        message: 'WiFi configured successfully',
        connectionTest: {
          success: true,
          ping: Math.floor(Math.random() * 50) + 10, // 10-60ms
          downloadSpeed: Math.floor(Math.random() * 50) + 10, // 10-60 Mbps
          uploadSpeed: Math.floor(Math.random() * 20) + 5, // 5-25 Mbps
        },
      };

      // En una implementación real, harías esto:
      // const response = await apiService.post(`${this.baseEndpoint}/${request.deviceId}/wifi-config`, request);
      // return response.data;

      return mockResponse;
    } catch (error) {
      logger.error('Error configuring WiFi:', error);
      
      return {
        success: false,
        message: 'Failed to configure WiFi. Please check your credentials and try again.',
      };
    }
  }

  /**
   * Obtener estado del robot
   */
  async getRobotStatus(deviceId: string): Promise<RobotStatus> {
    try {
      logger.debug('Getting robot status:', deviceId);

      // Simular llamada al backend
      await new Promise(resolve => setTimeout(resolve, 1000));

      // Simular respuesta del backend
      const mockStatus: RobotStatus = {
        deviceId,
        status: 'online',
        batteryLevel: Math.floor(Math.random() * 40) + 60, // 60-100%
        signalStrength: Math.floor(Math.random() * 30) + 70, // 70-100%
        lastCommand: 'wifi_configured',
        lastCommandTime: new Date().toISOString(),
        errors: [],
      };

      // En una implementación real, harías esto:
      // const response = await apiService.get(`${this.baseEndpoint}/${deviceId}/status`);
      // return response.data;

      return mockStatus;
    } catch (error) {
      logger.error('Error getting robot status:', error);
      
      return {
        deviceId,
        status: 'error',
        errors: ['Failed to get robot status'],
      };
    }
  }

  /**
   * Registrar un nuevo robot en el backend
   */
  async registerRobot(robotData: Partial<RobotDevice>): Promise<RobotDevice> {
    try {
      logger.debug('Registering new robot:', robotData);

      // Llamada real al backend
      const response = await apiService.post(`${this.baseEndpoint}`, robotData);
      
      return {
        id: response.data.id,
        name: response.data.name,
        model: response.data.model,
        serialNumber: response.data.serialNumber,
        firmwareVersion: response.data.firmwareVersion,
        bluetoothId: response.data.bluetoothId,
        status: response.data.status,
        lastSeen: response.data.lastSeen,
        createdAt: response.data.createdAt,
        updatedAt: response.data.updatedAt,
      };
    } catch (error) {
      logger.error('Error registering robot:', error);
      throw error;
    }
  }

  /**
   * Actualizar información del robot
   */
  async updateRobot(deviceId: string, updateData: Partial<RobotDevice>): Promise<RobotDevice> {
    try {
      logger.debug('Updating robot:', deviceId, updateData);

      // Simular llamada al backend
      await new Promise(resolve => setTimeout(resolve, 1000));

      // Simular respuesta del backend
      const mockUpdatedRobot: RobotDevice = {
        id: deviceId,
        name: updateData.name || 'NEBU Robot',
        model: updateData.model || 'NEBU-Robot-v1.0',
        serialNumber: updateData.serialNumber || 'NEBU-001-2024',
        firmwareVersion: updateData.firmwareVersion || '1.2.3',
        bluetoothId: updateData.bluetoothId || '',
        status: updateData.status || 'online',
        lastSeen: new Date().toISOString(),
        createdAt: new Date(Date.now() - 86400000).toISOString(),
        updatedAt: new Date().toISOString(),
      };

      // En una implementación real, harías esto:
      // const response = await apiService.put(`${this.baseEndpoint}/${deviceId}`, updateData);
      // return response.data;

      return mockUpdatedRobot;
    } catch (error) {
      logger.error('Error updating robot:', error);
      throw error;
    }
  }

  /**
   * Enviar comando al robot
   */
  async sendCommand(deviceId: string, command: string, parameters?: any): Promise<boolean> {
    try {
      logger.debug('Sending command to robot:', deviceId, command, parameters);

      // Simular llamada al backend
      await new Promise(resolve => setTimeout(resolve, 500));

      // Simular respuesta del backend
      const success = Math.random() > 0.1; // 90% success rate

      // En una implementación real, harías esto:
      // const response = await apiService.post(`${this.baseEndpoint}/${deviceId}/commands`, {
      //   command,
      //   parameters,
      //   timestamp: new Date().toISOString(),
      // });
      // return response.data.success;

      return success;
    } catch (error) {
      logger.error('Error sending command to robot:', error);
      return false;
    }
  }

  /**
   * Obtener lista de robots del usuario
   */
  async getUserRobots(): Promise<RobotDevice[]> {
    try {
      logger.debug('Getting user robots');

      // Llamada real al backend
      const response = await apiService.get(`${this.baseEndpoint}/user`);
      
      return response.data.map((device: any) => ({
        id: device.id,
        name: device.name,
        model: device.model,
        serialNumber: device.serialNumber,
        firmwareVersion: device.firmwareVersion,
        bluetoothId: device.bluetoothId,
        status: device.status,
        lastSeen: device.lastSeen,
        createdAt: device.createdAt,
        updatedAt: device.updatedAt,
      }));
    } catch (error) {
      logger.error('Error getting user robots:', error);
      return [];
    }
  }

  /**
   * Eliminar robot
   */
  async deleteRobot(deviceId: string): Promise<boolean> {
    try {
      logger.debug('Deleting robot:', deviceId);

      // Simular llamada al backend
      await new Promise(resolve => setTimeout(resolve, 1000));

      // En una implementación real, harías esto:
      // const response = await apiService.delete(`${this.baseEndpoint}/${deviceId}`);
      // return response.data.success;

      return true;
    } catch (error) {
      logger.error('Error deleting robot:', error);
      return false;
    }
  }
}

// Instancia singleton del servicio
export const robotService = new RobotService();
export default robotService;
